frb_codegen/src/tools.rs

```diff
@@ -0,0 +1,442 @@
+        let lock_file: PubspecLock = serde_yaml::from_str(&lock_file)
+            .map_err(|_| anyhow::Error::msg(format!("unable to parse {} in {}", filename, s)))?;
+        if lock_file
+            .packages
+            .contains_key(&DartToolchain::Flutter.to_string())
+        {
+            return Ok(DartRepository {
+                at: PathBuf::from(s),
+                toolchain: DartToolchain::Flutter,
+            });
+        }
+        Ok(DartRepository {
+            at: PathBuf::from(s),
+            toolchain: DartToolchain::Dart,
+        })
+    }
+}
+
+impl DartRepository {
+    /// check whether the toolchain is available from the CLI
+    pub(crate) fn toolchain_available(&self) -> bool {
+        self.toolchain.available()
+    }
+    /// check whether a package has been correctly specified in pubspec.yaml
+    pub(crate) fn has_specified(
+        &self,
+        package: &str,
+        manager: PackageManager,
+        requirement: &VersionReq,
+    ) -> anyhow::Result<()> {
+        let at = self.at.to_str().unwrap();
+        debug!("Checking presence of {} in {} at {}", package, manager, at);
+        let manifest_file = read_file(at, DartToolchain::manifest_filename())?;
+        let manifest_file: Pubspec = serde_yaml::from_str(&manifest_file).map_err(|_| {
+            anyhow::Error::msg(format!(
+                "unable to parse {} in {}",
+                DartToolchain::manifest_filename(),
+                at
+            ))
+        })?;
+        let deps = if manager == PackageManager::DevDependencies {
+            manifest_file.dev_dependencies.unwrap_or_default()
+        } else {
+            manifest_file.dependencies.unwrap_or_default()
+        };
+        deps.get(package).map(|_| ()).ok_or_else(|| {
+            anyhow::Error::new(Error::MissingDep {
+                name: package.to_string(),
+                manager,
+                requirement: requirement.to_string(),
+            })
+        })
+    }
+    /// check whether a package has been correctly pinned in pubspec.lock
+    pub(crate) fn has_installed(
+        &self,
+        package: &str,
+        manager: PackageManager,
+        requirement: &VersionReq,
+    ) -> anyhow::Result<()> {
+        let at = self.at.to_str().unwrap();
+        debug!("Checking presence of {} in {} at {}", package, manager, at);
+        let lock_file = read_file(at, DartToolchain::lock_filename())?;
+        let lock_file: PubspecLock = serde_yaml::from_str(&lock_file).map_err(|_| {
+            anyhow::Error::msg(format!(
+                "unable to parse {} in {}",
+                DartToolchain::lock_filename(),
+                at
+            ))
+        })?;
+        let dependency = lock_file.packages.get(package);
+        let version = match dependency {
+            Some(dependency) => {
+                let pm = dependency.installed_in();
+                if pm.is_none() || (pm.unwrap() != manager) {
```

Comment:
uber-nit: `pm != Some(manager)`

Reply:
damn I didn't even know it was possible to check a condition like this lol
done in 47037b7d40a6afe68d16be5fff38c3253e9fa7cb